File: app/src/main/java/com/simplertutorials/android/wheathograophy/data/api/ApiRepository.java
package com.simplertutorials.android.wheathograophy.data.api;

import com.simplertutorials.android.wheathograophy.BuildConfig;
import com.simplertutorials.android.wheathograophy.domain.ApiWeatherResponse;
import com.simplertutorials.android.wheathograophy.domain.City;

import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

public class ApiRepository {
    private static ApiRepository instance;


    private ApiRepository() {

    }

    public static ApiRepository getInstance() {
        if (instance == null)
            instance = new ApiRepository();
        return instance;
    }

    public Observable<ApiWeatherResponse> getWeatherInfo(ApiService apiService, City city) {
        return apiService.getWeather(city.getName(),
                BuildConfig.apiKey)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .toObservable();

    }
}

File: app/src/main/java/com/simplertutorials/android/wheathograophy/data/api/ApiRepository.kt
package com.simplertutorials.android.wheathograophy.data.api

import com.simplertutorials.android.wheathograophy.BuildConfig
import com.simplertutorials.android.wheathograophy.data.api.util.NetworkError
import com.simplertutorials.android.wheathograophy.data.api.util.Result
import com.simplertutorials.android.wheathograophy.domain.ApiWeatherResponse
import com.simplertutorials.android.wheathograophy.domain.City

class ApiRepository(
    private val apiService: ApiService
) {

    suspend fun getWeatherInfo(city: City): Result<ApiWeatherResponse, NetworkError> {
        return apiService.getWeather(
            city.name,
            BuildConfig.API_KEY
        )
    }
}
